added google sign-in feature

bryng001 opened the pull request:
**Summary**:
Added the google sign in feature using OAuth.
![image](https://github.com/user-attachments/assets/2510985c-d7c5-459a-bd79-6dbc7a62f7de)


**Description for the changelog**:
Added the google sign UI and provider in both frontend and backend

**Other info**:
For the google sign in feature, We need to add some env variables for Google Auth.
For example:
GOOGLE_CLIENT_ID=your-google-client-id
GOOGLE_CLIENT_SECRET=your-google-client-secret
GOOGLE_SCOPE=openid email profile
GOOGLE_REDIRECT_URI=http://localhost:3000/api/oauth/return

You can create a new project in Google Cloud Console and get the Google cloud credentials there.

Closes #1088

jgadsden commented:
we are a bit strict on the linting @bryng001 , if you want to correct that and rerun the tests

bryng001 commented:
> we are a bit strict on the linting @bryng001 , if you want to correct that and rerun the tests

Just fixed the lint issue,

ericfitz commented:
I think that you have to update https://github.com/OWASP/threat-dragon/blob/main/td.server/test/providers/index.spec.js to add the google.js provider in order to pass the server unit tests.  The tests are failing due to an unexpected provider.

bryng001 commented:
Just added the test code for the google sign in provider part,

jgadsden reviewed (changes requested):
thanks for doing this so quickly @bryng001 

did you want to add unit tests for the frontend? ideally in a file `td.vue/tests/unit/service/provider/github.provider.spec.js`
having said that, there are no unit tests for bitbucket or gitlab in the `td.vue/tests/unit/service/provider/` directory so this can be regarded as a 'nice to have'

bryng001 commented:
> thanks for doing this so quickly @bryng001
> 
> did you want to add unit tests for the frontend? ideally in a file `td.vue/tests/unit/service/provider/github.provider.spec.js` having said that, there are no unit tests for bitbucket or gitlab in the `td.vue/tests/unit/service/provider/` directory so this can be regarded as a 'nice to have'

Yes, Now I am adding unit tests for the frontend, also I think I can write the test code for the bitbucket and gitlab as well.
Currently the e2e smokes has failed for my PR, can you please let me know what I should update to pass this test?
Also the second feature is to integrate google drive API, is it ok if it would work like in local using Google Drive space?
Please let me know the details.

bryng001 commented:
Hi Eric, here is an update of the progress:
9/30/24
What I did:
Checked the entire app logic, especially github integration part.
Did some research to integrate Google Drive API using access_token that can get from the Google sign-in.
Wrote some unit test code for the Google sign-in feature.
What I will do today:
Will try to integrate the Google Drive API and test some basic features just like getting folder names from Google Drive.
Roadblocks: No
Estimated completion: 40%
Project delivery status: Yellow



I don't have any questions so far. Do let me know you if have additional requests. Thanks!

jgadsden commented:
yes, good point @bryng001 , the smoke tests fail because you are not privileged enough to run the docker container in the pipeline
I will create an issue for it and will try to fix this for you

> Currently the e2e smokes has failed for my PR, can you please let me know what I should update to pass this test?

bryng001 commented:
Hi Eric, here is an update of the progress:
10/2/24
What I did:
Integrated Google Drive API and implemented same logic with the Github provider.
Added unit test code for new features.
What I will do today:
Will review all the features what I implemented again.
Roadblocks: No
Estimated completion: 95%
Project delivery status: Green

I think you can test this Google Drive feature after merged this PR and deployed to prod.

ericfitz commented on td.server/src/env/Google.js:
```diff
@@ -0,0 +1,25 @@
+            { key: 'CLIENT_ID', required: false },
+            { key: 'CLIENT_SECRET', required: false },
+            { key: 'SCOPE', required: false, defaultValue: 'openid email profile' },
+            { key: 'REDIRECT_URI', required: false, defaultValue: 'http://localhost:8080/oauth-return' },
```
Out of curiosity, why is the port here different than in example.env?

bryng001 replied:
Oh, I think I have write the sample redirect url from the Google Drive API doc.
So is it ok if I update it with http://localhost:3000/api/oauth/return ?

ericfitz replied:
Sure!

bryng001 replied:
Hi @ericfitz .
Just updated the default env variable for Google Redirect URL. Also this PR has passed all the checks, Please check.

bryng001 replied:
Hello @ericfitz .

Also, we can end the current contract in Upwork and leave reviews for each other, It'd be very helpful for me.
As you can see, my account is new so it'd better if I have more good reviews from clients.
There is two kind of score in Upwork when you end the contract, Please give me private score 10/10, and public score 5/5 and your reviews for my work.
If you need my help again, we can start a new contract there, in this way I can have more reviews, Thank you very much.

jgadsden commented:
@bryng001 I have fixed up the pull request pipeline so that all the stages should pass for you
(before it was only working ro a relatively privileged account)
do you want to rebase to main to pick up the changes to pipelines / workflows?
no problem if you skip this, as it is only for the e2e tests that I can extend later on

bryng001 commented:
> @bryng001 I have fixed up the pull request pipeline so that all the stages should pass for you (before it was only working for a relatively privileged account) do you want to rebase to main to pick up the changes to pipelines / workflows? no problem if you skip this, as it is only for the e2e tests that I can extend later on

Okay, Sure, I will rebase it and update the PR again. Thanks.

bryng001 commented:
Hi, @jgadsden 
I have just completed the rebase and passed all the test, Please check.

ericfitz commented:
Thanks for all the hard work, Brian!  And Jon, thank you for all your help!

ericfitz commented:
Did I just make a mistake in closing this?

bryng001 commented:
> Did I just make a mistake in closing this?

This PR has not merged yet, Maybe you can reopen this PR in your end or Jon can do this.

ericfitz commented:
Hey @jgadsden would you please do the merge?

Thank you!

jgadsden commented:
> Hey @jgadsden would you please do the merge?

Thanks for this @bryng001 and @ericfitz , I will try it out and as I go along create a new documentation page for it - along the lines of https://owasp.org/www-project-threat-dragon/docs-2/gitlab-repo/
I should then be able to merge it to main

jgadsden reviewed (approved):
excellent to have this @bryng001 , thanks for all your work
merging this to main
